File: telegram-wife-assistant/store.js
// =====================================================================
// store.js — المخزن المحلي (ملف JSON واحد).
// بيحتفظ بـ:
//   - feedback[]:        سجل كل تفاعل (تاريخ، خانة، مواضيع، اختيار، نص نهائي)
//   - styleExamples[]:   آخر ≤30 رسالة انت اخترتها/عدّلتها = "ملف الأسلوب"
//   - themeWeights{}:    أوزان المواضيع للترجيح (المختار يزيد، المتجاهَل يقل)
//   - lastSentPerSlot{}: آخر يوم اتبعت فيه كل خانة (عشان ما نكررش في نفس اليوم)
//
// كله ملف بسيط، سهل تراجعه وتاخد منه نسخة احتياطية بإيدك.
// =====================================================================

const fs = require('fs');
const path = require('path');
const config = require('./config');
const { todayISO, daysAgoISO } = require('./util');

// مسار المخزن. افتراضياً data/store.json جنب المشروع، ويمكن تغييره عبر
// متغيّر البيئة WIFE_ASSISTANT_STORE_FILE (بنستخدمه في الاختبارات عشان ما
// نلمسش بياناتك الحقيقية).
const STORE_PATH = process.env.WIFE_ASSISTANT_STORE_FILE
  ? path.resolve(process.env.WIFE_ASSISTANT_STORE_FILE)
  : path.join(__dirname, 'data', 'store.json');
const DATA_DIR = path.dirname(STORE_PATH);

// الشكل الافتراضي لأول تشغيل.
function defaultStore() {
  const themeWeights = {};
  for (const t of config.themes) themeWeights[t] = 1; // وزن مبدئي متساوي
  return {
    feedback: [], // [{ id, date, slot, themesShown, choice, finalText }]
    styleExamples: [], // [{ text, theme, date }]
    themeWeights, // { 'امتنان': 1.4, ... }
    lastSentPerSlot: {}, // { morning: '2026-06-21', evening: '...' }
    review: { lastWeeklyAt: null }, // حالة المراجعة الأسبوعية
  };
}

// قراءة المخزن (يصلّح نفسه لو الملف ناقص حقول).
function read() {
  try {
    if (!fs.existsSync(STORE_PATH)) return defaultStore();
    const raw = fs.readFileSync(STORE_PATH, 'utf8');
    const parsed = JSON.parse(raw);
    // دمج مع الافتراضي عشان أي حقل جديد ما يبوّظش الملفات القديمة.
    return { ...defaultStore(), ...parsed };
  } catch (err) {
    // ملف بايظ ≠ نبدأ من الصفر بصمت: بنحطه جنباً عشان تقدر تستعيده يدوياً.
    try {
      const quarantine = `${STORE_PATH}.corrupt-${Date.now()}`;
      fs.renameSync(STORE_PATH, quarantine);
      console.error(`⚠️ المخزن تالف — اتنقل إلى ${quarantine} وهنبدأ بنسخة جديدة:`, err.message);
    } catch (renameErr) {
      console.error('⚠️ تعذّر قراءة المخزن، هنبدأ بنسخة جديدة:', err.message);
    }
    return defaultStore();
  }
}

// كتابة المخزن على القرص.
function write(store) {
  if (!fs.existsSync(DATA_DIR)) fs.mkdirSync(DATA_DIR, { recursive: true });
  // كتابة ذرّية: ملف مؤقّت ثم rename، عشان انقطاع الكهرباء ما يبوّظش المخزن.
  const tmp = `${STORE_PATH}.tmp`;
  fs.writeFileSync(tmp, JSON.stringify(store, null, 2), 'utf8');
  fs.renameSync(tmp, STORE_PATH);
}

// تاريخ النهاردة YYYY-MM-DD (مصدره util عشان يبقى موحّد مع باقي المشروع).
const today = todayISO;

// ---- ثبات الحالة: هل اتبعت الخانة دي النهاردة قبل كده؟ ----
function wasSlotSentToday(slot) {
  const store = read();
  return store.lastSentPerSlot[slot] === today();
}

function markSlotSentToday(slot) {
  const store = read();
  store.lastSentPerSlot[slot] = today();
  write(store);
}

// ---- ملف الأسلوب: إضافة مثال جديد (مع سقف 30، الأقدم يخرج) ----
function addStyleExample(text, theme) {
  const store = read();
  store.styleExamples.push({ text: text.trim(), theme: theme || null, date: today() });
  // قص على آخر styleExamplesMax مثال.
  const max = config.styleExamplesMax;
  if (store.styleExamples.length > max) {
    store.styleExamples = store.styleExamples.slice(-max);
  }
  write(store);
}

function getStyleExamples() {
  return read().styleExamples;
}

// ---- ترجيح المواضيع ----
// المختار: نزوّد وزنه. المتجاهَل: نقلّل وزنه (مع حد أدنى عشان ما يختفيش خالص).
function bumpThemeWeight(theme, delta) {
  if (!theme) return;
  const store = read();
  const current = store.themeWeights[theme] ?? 1;
  let next = current + delta;
  if (next < 0.2) next = 0.2; // حد أدنى: الموضوع يفضل وارد بنسبة صغيرة
  if (next > 5) next = 5; // حد أقصى: ما يطغاش على الباقي
  store.themeWeights[theme] = Number(next.toFixed(3));
  write(store);
}

function getThemeWeights() {
  return read().themeWeights;
}

// ---- المواضيع المستخدمة في آخر N يوم (للتنويع) ----
function recentThemes(days) {
  const store = read();
  const cutoffStr = daysAgoISO(days);
  const themes = new Set();
  for (const fb of store.feedback) {
    if (fb.date >= cutoffStr && Array.isArray(fb.themesShown)) {
      for (const t of fb.themesShown) themes.add(t);
    }
  }
  return [...themes];
}

// ---- تسجيل تغذية راجعة ----
function addFeedback(entry) {
  const store = read();
  store.feedback.push({ id: Date.now(), date: today(), ...entry });
  write(store);
}

function getFeedback() {
  return read().feedback;
}

// ---- المراجعة الأسبوعية: ختم آخر مرة اتبعت ----
function markWeeklyReviewSent() {
  const store = read();
  store.review.lastWeeklyAt = new Date().toISOString();
  write(store);
}

// ---- /reset: تصفير التعلّم بالكامل (الأمثلة + الأوزان + التغذية الراجعة) ----
function resetLearning() {
  const fresh = defaultStore();
  // نحافظ على lastSentPerSlot عشان ما نبعتش مرتين في نفس اليوم بعد الريسيت.
  const store = read();
  fresh.lastSentPerSlot = store.lastSentPerSlot;
  write(fresh);
}

module.exports = {
  read,
  write,
  today,
  wasSlotSentToday,
  markSlotSentToday,
  addStyleExample,
  getStyleExamples,
  bumpThemeWeight,
  getThemeWeights,
  recentThemes,
  addFeedback,
  getFeedback,
  markWeeklyReviewSent,
  resetLearning,
  STORE_PATH,
};

File: android-wife-assistant/app/src/main/java/com/wifeassistant/data/Store.kt
package com.wifeassistant.data

import android.content.Context
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File

// المخزن المحلي (JSON) - نظير store.js. بيحفظ التعلّم وثبات الحالة.
class Store(context: Context) {
    private val file = File(context.filesDir, "store.json")
    private val json = Json { prettyPrint = true; ignoreUnknownKeys = true; encodeDefaults = true }

    @Synchronized
    fun read(): AppData {
        if (!file.exists()) return defaultData()
        return runCatching { json.decodeFromString<AppData>(file.readText()) }.getOrElse {
            // ملف تالف ≠ مسح التعلّم بصمت: بنحجره جنباً ونبدأ نظيف.
            runCatching { file.renameTo(File(file.parentFile, "store.json.corrupt-${System.currentTimeMillis()}")) }
            defaultData()
        }
    }

    @Synchronized
    fun write(data: AppData) {
        // كتابة ذرّية: ملف مؤقّت ثم rename، عشان انقطاع الكتابة ما يبوّظش المخزن.
        val tmp = File(file.parentFile, "store.json.tmp")
        tmp.writeText(json.encodeToString(data))
        if (!tmp.renameTo(file)) {
            file.delete()
            tmp.renameTo(file)
        }
    }

    private fun defaultData(): AppData {
        val weights = AppConstants.THEMES.associateWith { 1.0 }.toMutableMap()
        return AppData(themeWeights = weights)
    }

    // ---- ملف الأسلوب (آخر 30 لكل شخص على حدة) ----
    fun styleExamples(recipientId: String): List<StyleExample> =
        read().styleExamples.filter { it.recipientId == recipientId }

    fun styleExamplesCount(): Int = read().styleExamples.size

    @Synchronized
    fun addStyleExample(text: String, theme: String?, recipientId: String) {
        val d = read()
        d.styleExamples.add(StyleExample(text.trim(), theme, DateUtil.todayISO(), recipientId))
        // السقف لكل شخص لوحده (الأقدم لنفس الشخص يخرج).
        while (d.styleExamples.count { it.recipientId == recipientId } > AppConstants.STYLE_EXAMPLES_MAX) {
            val idx = d.styleExamples.indexOfFirst { it.recipientId == recipientId }
            if (idx >= 0) d.styleExamples.removeAt(idx) else break
        }
        write(d)
    }

    // ---- أوزان المواضيع (ترجيح) ----
    fun themeWeights(): Map<String, Double> = read().themeWeights

    @Synchronized
    fun bumpThemeWeight(theme: String?, delta: Double) {
        if (theme == null) return
        val d = read()
        var next = (d.themeWeights[theme] ?: 1.0) + delta
        next = next.coerceIn(0.2, 5.0)
        d.themeWeights[theme] = next
        write(d)
    }

    fun recentThemes(days: Long): Set<String> {
        val cutoff = DateUtil.daysAgoISO(days)
        return read().feedback.filter { it.date >= cutoff }.flatMap { it.themesShown }.toSet()
    }

    // ---- تغذية راجعة ----
    @Synchronized
    fun addFeedback(fb: Feedback) {
        val d = read(); d.feedback.add(fb); write(d)
    }

    fun feedback(): List<Feedback> = read().feedback

    // حذف عنصر من السجل (بمطابقة التاريخ والنص).
    @Synchronized
    fun deleteHistory(date: String, text: String) {
        val d = read()
        d.feedback.removeAll { it.date == date && it.finalText == text }
        write(d)
    }

    // ---- المفضّلة ----
    fun favorites(): List<String> = read().favorites
    fun isFavorite(text: String): Boolean = read().favorites.contains(text)

    @Synchronized
    fun toggleFavorite(text: String) {
        val d = read()
        if (!d.favorites.remove(text)) d.favorites.add(text)
        write(d)
    }

    // ---- ثبات الحالة (مش نفس الخانة مرتين/يوم) ----
    fun wasSlotSentToday(slot: String): Boolean = read().lastSentPerSlot[slot] == DateUtil.todayISO()

    @Synchronized
    fun markSlotSentToday(slot: String) {
        val d = read(); d.lastSentPerSlot[slot] = DateUtil.todayISO(); write(d)
    }

    // ---- تذكيرات التواصل ("بقالك فترة ما كلّمت فلان") ----
    @Synchronized
    fun markContacted(recipientId: String) {
        if (recipientId.isBlank()) return
        val d = read(); d.lastContactedPerRecipient[recipientId] = DateUtil.todayISO(); write(d)
    }

    // عدد الأيام من آخر تواصل، أو null لو مفيش تواصل قبل كده.
    fun daysSinceContact(recipientId: String): Long? {
        val last = read().lastContactedPerRecipient[recipientId] ?: return null
        return runCatching {
            java.time.temporal.ChronoUnit.DAYS.between(
                java.time.LocalDate.parse(last),
                DateUtil.today(),
            )
        }.getOrNull()
    }

    // ---- الجولة المعلّقة (بين الإشعار والواجهة) ----
    fun getPending(): PendingRound? = read().pending

    @Synchronized
    fun setPending(p: PendingRound) {
        val d = read(); d.pending = p; write(d)
    }

    @Synchronized
    fun clearPending() {
        val d = read(); d.pending = null; write(d)
    }

    // ---- تصفير التعلّم (بيحافظ على lastSentPerSlot) ----
    @Synchronized
    fun resetLearning() {
        val old = read()
        val fresh = defaultData()
        fresh.lastSentPerSlot.putAll(old.lastSentPerSlot)
        write(fresh)
    }
}
